modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryDisconnectingTest.java

```diff
@@ -0,0 +1,147 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi.discovery.tcp;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.internal.managers.discovery.CustomEventListener;
+import org.apache.ignite.internal.managers.discovery.CustomMessageWrapper;
+import org.apache.ignite.internal.managers.discovery.DiscoCache;
+import org.apache.ignite.internal.managers.discovery.DiscoveryCustomMessage;
+import org.apache.ignite.internal.managers.discovery.GridDiscoveryManager;
+import org.apache.ignite.internal.processors.affinity.AffinityTopologyVersion;
+import org.apache.ignite.internal.util.lang.GridAbsPredicate;
+import org.apache.ignite.internal.util.typedef.G;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.spi.discovery.tcp.internal.TcpDiscoverySpiState;
+import org.apache.ignite.spi.discovery.tcp.messages.TcpDiscoveryEnsureDelivery;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.jetbrains.annotations.Nullable;
+import org.mockito.internal.util.reflection.Whitebox;
+
+/**
+ *
+ */
+public class TcpDiscoveryDisconnectingTest extends GridCommonAbstractTest {
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testCustomMessageWhileSpiIsStopping() throws Exception {
+        List<DummyCustomDiscoveryMessage> messages = new CopyOnWriteArrayList<>();
+
+        startGrids(3);
+
+        CustomEventListener<DummyCustomDiscoveryMessage> listener = new CustomEventListener<DummyCustomDiscoveryMessage>() {
+            @Override public void onCustomEvent(
+                AffinityTopologyVersion topVer,
+                ClusterNode snd,
+                DummyCustomDiscoveryMessage msg) {
+                messages.add(msg);
+            }
+        };
+
+        G.allGrids().forEach(new Consumer<Ignite>() {
+            @Override public void accept(Ignite ignite) {
+                ((IgniteEx)ignite)
+                    .context()
+                    .discovery()
+                    .setCustomEventListener(DummyCustomDiscoveryMessage.class, listener);
+            }
+        });
+
+        grid(0).cluster().active(true);
+
+        TcpDiscoverySpi coordDisco = (TcpDiscoverySpi)grid(0).configuration().getDiscoverySpi();
+
+        TcpDiscoverySpi nodeDisco = (TcpDiscoverySpi)grid(1).configuration().getDiscoverySpi();
+
+        TcpDiscoveryImpl impl = (TcpDiscoveryImpl)Whitebox.getInternalState(nodeDisco, "impl");
+        Whitebox.setInternalState(impl, "spiState", TcpDiscoverySpiState.DISCONNECTING);
```

Comment:
This is not a unit test. You can't just change a field value via reflection.
Try to achieve the same effect by calling API methods.